src/main/java/org/craftercms/studio/impl/v2/service/dashboard/DashboardServiceImpl.java

```diff
@@ -309,17 +334,23 @@ public List<DashboardPublishingPackage> getPublishingHistory(
                 dateTo, offset, limit);
     }
 
+    @Override
+    @HasPermission(type = DefaultPermission.class, action = PERMISSION_CONTENT_READ)
+    public int getPublishingHistoryDetailTotalItems(@ProtectedResourceId(SITE_ID_RESOURCE_ID) String siteId,
+                                                    String publishingPackageId) throws SiteNotFoundException {
+        siteService.checkSiteExists(siteId);
+        return publishServiceInternal.getPublishingHistoryDetailTotalItems(siteId, publishingPackageId);
+    }
+
     @Override
     @HasPermission(type = DefaultPermission.class, action = PERMISSION_CONTENT_READ)
     public List<SandboxItem> getPublishingHistoryDetail(@ProtectedResourceId(SITE_ID_RESOURCE_ID) String siteId,
-                                                        String publishingPackageId)
+                                                        String publishingPackageId, int offset, int limit)
             throws UserNotFoundException, ServiceLayerException {
         siteService.checkSiteExists(siteId);
-        // TODO: Try to do a single query
-        var publishingPackageDetails =
-                publishServiceInternal.getPublishingPackageDetails(siteId, publishingPackageId);
-        var paths = publishingPackageDetails.getItems().stream()
-                .map(PublishingPackageDetails.PublishingPackageItem::getPath)
+        var packageDetail = publishServiceInternal.getPublishingHistoryDetail(siteId, publishingPackageId, offset, limit);
+        var paths = packageDetail.stream()
+                .map(PublishRequest::getPath)
                 .collect(toList());
         if (isEmpty(paths)) {
```

Comment:
Can this `isEmpty` check move up before the stream (check if package detail is empty)?

Reply:
Fixed. Thank you!